[libc++] Avoid re-declaring ::lgamma_r from std::binomial_distribution

ldionne opened the pull request:
Libc++ currently redeclares ::lgamma_r on platforms that provide it. This causes issues when building with modules, and redeclaring functions provided by another library (here the C library) is bad hygiene.

Instead, use an asm declaration to call the right function without having to redeclare it.

ldionne committed:
[libc++] Avoid redeclaring lgamma_r

This causes issues when building with modules, and redeclaring functions
provided by another library (here the C library) is bad hygiene. Instead,
use <math.h> to access the declaration provided by the system.

As a drive-by, this also starts using ::lgamma_r instead of ::lgamma when
building on top of LLVM-libc. We didn't do that previously due to a
declration conflict, but using the _r version when we can should only
make things thread safe.

philnik777 reviewed (commented):
Is `lgamma_r` unconditionally provided? I think it would be really nice if we could get a private implementation from LLVM libc so that we can be thread-safe on every platform.

philnik777 commented:
> > Is `lgamma_r` unconditionally provided? I think it would be really nice if we could get a private implementation from LLVM libc so that we can be thread-safe on every platform.
> 
> I do think it is unconditionally provided by llvm-libc. What you're thinking about is some kind of hand-in-hand like thing but just for that function?

Yeah. We could just define our own `__lgamma_r` or something so we can have it on all platforms. That would also remove the need to expose extensions through `_REENTRANT` or anything like that. Actually, we could just always define our own `__lgamma_r` and for back-deployment we can add an `__asm__("lgamma_r")`. Apple platforms are the only ones where we provide any back-deployment guarantees, so that should be maintainable, and eventually we can remove that workaround as well.

yuxuanchen1997 commented:
> > Is `lgamma_r` unconditionally provided? I think it would be really nice if we could get a private implementation from LLVM libc so that we can be thread-safe on every platform.
> 
> I do think it is unconditionally provided by llvm-libc. What you're thinking about is some kind of hand-in-hand like thing but just for that function?

From the #libcxx channel, someone mentioned that LLVM libc doesn't implement `lgamma_r` function. Maybe this suggestion is a longer shot than expected?

yuxuanchen1997 commented on libcxx/include/__random/binomial_distribution.h:
```diff
@@ -15,6 +15,15 @@
+#endif
+#include <math.h> // for ::lgamma_r
+#ifdef _LIBCPP_MUST_UNDEFINE_REENTRANT
+#  undef _REENTRANT
```
undef _LIBCPP_MUST_UNDEFINE_REENTRANT as well?

jhuber6 commented:
> > Is `lgamma_r` unconditionally provided? I think it would be really nice if we could get a private implementation from LLVM libc so that we can be thread-safe on every platform.
> 
> I do think it is unconditionally provided by llvm-libc. What you're thinking about is some kind of hand-in-hand like thing but just for that function?

I don't think we have a single generic implementation in llvm-libc yet, maybe @lntue has something.

ldionne committed:
Handle ABIlist

ldionne committed:
Link against -lm

philnik777 commented on libcxx/src/math.cpp:
```diff
@@ -0,0 +1,26 @@
+_LIBCPP_BEGIN_NAMESPACE_STD
+namespace __math {
+
+_LIBCPP_EXPORTED_FROM_ABI __lgamma_result __lgamma_thread_safe_impl(double __d) noexcept {
```
The `_LIBCPP_EXPORTED_FROM_ABI` shouldn't be required here.

philnik777 replied:
I have to say I completely disagree. IMO this significantly hurts readability. We should instead just have non-_Uglified names for library-internal functions (which itself helps with readability in general). Spotting exported ones is trivial in that case. Removing the attribute here also ensures that the correct declarations are visible at the definition, which helps avoid declaration mismatches.

ldionne replied:
I'm fine with removing it, but let's try to clean up other uses in `.cpp` files then.

philnik777 commented on libcxx/include/__math/gamma.h:
```diff
@@ -55,6 +55,31 @@ inline _LIBCPP_HIDE_FROM_ABI double tgamma(_A1 __x) _NOEXCEPT {
+#if _LIBCPP_AVAILABILITY_HAS_THREAD_SAFE_LGAMMA
+_LIBCPP_EXPORTED_FROM_ABI __lgamma_result __lgamma_thread_safe_impl(double) _NOEXCEPT;
+
+inline _LIBCPP_HIDE_FROM_ABI __lgamma_result __lgamma_thread_safe(double __d) _NOEXCEPT {
+  return __math::__lgamma_thread_safe_impl(__d);
+}
```
What's the reason for having another wrapper here?

ldionne replied:
I wanted to avoid having a function that is sometimes defined inline and sometimes defined in the dylib, since I think it's more difficult to reason about ODR and ABI-wise.

philnik777 commented on libcxx/include/__math/gamma.h:
```diff
@@ -55,6 +55,31 @@ inline _LIBCPP_HIDE_FROM_ABI double tgamma(_A1 __x) _NOEXCEPT {
 
+// __lgamma_r
+
+struct __lgamma_result {
+  double __result;
+  int __sign;
+};
```
We're never using the `__sign`. Should we remove that part? @lntue do you think this makes any difference in the implementation?

philnik777 commented on libcxx/src/math.cpp:
```diff
@@ -0,0 +1,26 @@
+#include <math.h> // for lgamma_r
+
+_LIBCPP_BEGIN_NAMESPACE_STD
+namespace __math {
```
Do we really want to enshrine the `__math` namespace forever in our ABI?

ldionne committed:
Workaround broken flag detection on GCC

ldionne committed:
Remove EXPORTED_FROM_ABI

ldionne committed:
Move out of __math namespace

ldionne committed:
Merge branch 'main' into review/dont-redeclare-lgamma_r

ldionne committed:
Update for LLVM 23

ldionne committed:
Try fixing the declaration of lgamma_r on mach-o

ldionne committed:
Handle MSVC

ldionne committed:
Don't define a new function in the dylib

Instead, just use an `asm` declaration to call the right symbol
when it is available. This reduces ABI lock in and provides the
same benefits.

ldionne committed:
Merge branch 'main' into review/dont-redeclare-lgamma_r

ldionne committed:
Remove availability

philnik777 reviewed (approved):
I'd still like our own version at some point, but this is definitely an improvement over the status quo, so I'm happy to do it like this for now.
